pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,437 @@
+		}
+		client := r.registryClients[ctx]
+		if err := client.Get(r.ctx, types.NamespacedName{Name: decoded.Namespace}, &corev1.Namespace{}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("failed to check if namespace %s exists: %w", decoded.Namespace, err)
+			}
+			if err := client.Create(r.ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: decoded.Namespace}}); err != nil && !apierrors.IsAlreadyExists(err) {
+				return fmt.Errorf("failed to create namespace %s: %w", decoded.Namespace, err)
+			}
+		}
+
+		imageStreamImport := &imagev1.ImageStreamImport{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: decoded.Namespace,
+				Name:      imageStreamName,
+			},
+			Spec: imagev1.ImageStreamImportSpec{
+				Import: true,
+				Images: []imagev1.ImageImportSpec{{
+					From: corev1.ObjectReference{
+						Kind: "DockerImage",
+						Name: publicURLForImage(sourceImageStreamTag.Image.DockerImageReference),
+					},
+					To: &corev1.LocalObjectReference{Name: imageTag},
+					ReferencePolicy: imagev1.TagReferencePolicy{
+						Type: imagev1.LocalTagReferencePolicy,
+					},
+				}},
+			},
+		}
+
+		r.log.Debugf("creating import %v on cluster %s", imageStreamImport, ctx)
+		if r.dryRun {
+			continue
+		}
+
+		if err := r.ensureCIOperatorRoleBinding(decoded.Namespace, client, log); err != nil {
+			return fmt.Errorf("failed to ensure rolebinding: %w", err)
+		}
+		if err := r.ensureCIOperatorRole(decoded.Namespace, client, log); err != nil {
+			return fmt.Errorf("failed to ensure role: %w", err)
+		}
+		if err := r.ensureImageStream(sourceImageStream, client, log); err != nil {
+			return fmt.Errorf("failed to ensure imagestream: %w", err)
+		}
+
+		if isTag != nil && isTag.Image.Name == sourceImageStreamTag.Image.Name {
+			log.Debug("ImageStreamTag is current")
+			return nil
+		}
+
+		if err := r.ensureImagePullSecret(decoded.Namespace, client, log); err != nil {
+			return fmt.Errorf("failed to ensure imagePullSecret: %w", err)
+		}
+
+		// ImageStreamImport is not an ordinary api but a virtual one that does the import synchronously
+		if err := client.Create(r.ctx, imageStreamImport); err != nil {
+			return fmt.Errorf("failed to import Image: %w", err)
+		}
+
+		// This should never be needed, but we shouldn't panic if the server screws up
+		if imageStreamImport.Status.Images == nil {
+			imageStreamImport.Status.Images = []imagev1.ImageImportStatus{{}}
+		}
+		if imageStreamImport.Status.Images[0].Image == nil {
+			return fmt.Errorf("imageStreamImport did not succeed: reason: %s, message: %s", imageStreamImport.Status.Images[0].Status.Reason, imageStreamImport.Status.Images[0].Status.Message)
+		}
+
+		log.Debug("Imported successfully")
+	}
+	return nil
+}
+
+func findNewest(isTags map[string]*imagev1.ImageStreamTag) (string, error) {
+	resultContext := ""
+	t := metav1.Now()
+	for ctx, isTag := range isTags {
+		if isTag == nil {
+			continue
+		}
+		if isTag.Image.CreationTimestamp.Before(&t) {
```

Comment:
If `creationTimestamp` is `before now()` its the `sourceContext`? This condition is always true. You need to compare the creation timestamps of the two, not compare against `now()`

Reply:
Updated.
Will add unit tests for this function.

Reply:
Done.